compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -1609,11 +1618,39 @@ trait Implicits:
             validateOrdering(ord)
             throw ex
 
-      val result = rank(sort(eligible), NoMatchingImplicitsFailure, Nil)
-      for (critical, msg) <- priorityChangeWarnings do
-        if result.found.exists(critical.contains(_)) then
-          report.warning(msg, srcPos)
-      result
+      val sorted = sort(eligible)
+      val res = sorted match
+        case first :: rest =>
+          val firstIsImplicit = first.ref.symbol.is(Implicit)
+          if rest.exists(_.ref.symbol.is(Implicit) != firstIsImplicit) then
+            // Mixture of implicits and givens
+            // Rank implicits first, then, if there is a given that it better than the best implicit(s)
+            // switch over to givens.
+            val (sortedImplicits, sortedGivens) = sorted.partition(_.ref.symbol.is(Implicit))
+            val implicitResult = rank(sortedImplicits, NoMatchingImplicitsFailure, Nil)
+            rank(sortedGivens, implicitResult, Nil)
+          else
+            rank(sorted, NoMatchingImplicitsFailure, Nil)
+        case _ =>
+          NoMatchingImplicitsFailure
+
+      // Issue all priority change warnings that can affect the result
+      val shownWarnings = priorityChangeWarnings.toList.collect:
+        case (critical, msg) if res.found.exists(critical.contains(_)) =>
+          msg
+      res match
+        case res: SearchFailure =>
+          res.reason match
+            case ambi: AmbiguousImplicits =>
+              // Make warnings part of error message because otherwise they are suppressed when
+              // the error is emitted.
+              ambi.priorityChangeWarnings = shownWarnings
+            case _ =>
+        case _ =>
+      for msg <- shownWarnings do
+        report.warning(msg, srcPos)
```

Comment:
I think this suffers from a combination of issues similar to the prior problems we have fixed:
- the errors swallowing the warnings about changes in resolution 
- the non-deterministic order of eligible alternatives
which may be causing the last error in the CI for `tests/warn/i15264.scala`.

The `report.warning` for multiple changes at the same source position may end up keeping a single warning for the one that happens to be reported last in `shownWarnings`.

Reply:
Yes I think that's it. I believe we should just drop the check file.